File: app/src/main/java/com/kakeibo/ui/MainActivity.java
package com.kakeibo.ui;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.viewpager.widget.ViewPager;

import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.Purchase;
import com.firebase.ui.auth.AuthUI;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.tabs.TabLayout;
import com.google.firebase.auth.FirebaseUser;
import com.kakeibo.Constants;
import com.kakeibo.R;
import com.kakeibo.SubApp;
import com.kakeibo.billing.BillingClientLifecycle;
import com.kakeibo.settings.SettingsCompatActivity;
import com.kakeibo.ui.search.TabFragment3;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";
    private static final int RC_SIGN_IN = 0;

    private SmartFragmentStatePagerAdapter _myFragmentPagerAdapter;
    private ViewPager _viewPager;

    private static FloatingActionButton fabStart;
    private static FloatingActionButton fabEnd;

    private BillingClientLifecycle _billingClientLifecycle;

    private FirebaseUserViewModel _authenticationViewModel;
    private BillingViewModel _billingViewModel;
    private SubscriptionStatusViewModel _subscriptionViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));

        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        _myFragmentPagerAdapter = new SmartPagerAdapter(getSupportFragmentManager());
        // Set up the ViewPager with the sections adapter.
        _viewPager = findViewById(R.id.viewpager);
        TabLayout tabs = findViewById(R.id.tabs);
        _viewPager.setAdapter(_myFragmentPagerAdapter);
        _viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabs));
        tabs.addOnTabSelectedListener(new TabLayout.ViewPagerOnTabSelectedListener(_viewPager));

        _authenticationViewModel = ViewModelProviders.of(this).get(FirebaseUserViewModel.class);
        _billingViewModel = ViewModelProviders.of(this).get(BillingViewModel.class);
        _subscriptionViewModel = ViewModelProviders.of(this).get(SubscriptionStatusViewModel.class);

        _billingClientLifecycle = ((SubApp) getApplication()).getBillingClientLifecycle();
        getLifecycle().addObserver(_billingClientLifecycle);

        // Register purchases when they change.
        _billingClientLifecycle.purchaseUpdateEvent.observe(this, new Observer<List<Purchase>>() {
            @Override
            public void onChanged(List<Purchase> purchases) {
                if (purchases != null) {
                    registerPurchases(purchases);
                }
            }
        });

        // Launch billing flow when user clicks button to buy something.
        _billingViewModel.buyEvent.observe(this, new Observer<BillingFlowParams>() {
            @Override
            public void onChanged(BillingFlowParams billingFlowParams) {
                if (billingFlowParams != null) {
                    _billingClientLifecycle
                            .launchBillingFlow(MainActivity.this, billingFlowParams);
                }
            }
        });

        // Open the Play Store when event is triggered.
        _billingViewModel.openPlayStoreSubscriptionsEvent.observe(this, new Observer<String>() {
            @Override
            public void onChanged(String sku) {
                Log.i(TAG, "Viewing subscriptions on the Google Play Store");
                String url;
                if (sku == null) {
                    // If the SKU is not specified, just open the Google Play subscriptions URL.
                    url = Constants.PLAY_STORE_SUBSCRIPTION_URL;
                } else {
                    // If the SKU is specified, open the deeplink for this SKU on Google Play.
                    url = String.format(Constants.PLAY_STORE_SUBSCRIPTION_DEEPLINK_URL,
                            sku, getApplicationContext().getPackageName());
                }
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                startActivity(intent);
            }
        });

        // Update authentication UI.
        final Observer<FirebaseUser> fireaseUserObserver = new Observer<FirebaseUser>() {
            @Override
            public void onChanged(@Nullable final FirebaseUser firebaseUser) {
                invalidateOptionsMenu();
                if (firebaseUser == null) {
//                    triggerSignIn();
                } else {
                    Log.d(TAG, "Current user: "
                            + firebaseUser.getEmail() + " " + firebaseUser.getDisplayName());
                }
            }
        };
        _authenticationViewModel.firebaseUser.observe(this, fireaseUserObserver);

        // Update subscription information when user changes.
        _authenticationViewModel.userChangeEvent.observe(this, new Observer<Void>() {
            @Override
            public void onChanged(Void aVoid) {
                _subscriptionViewModel.userChanged();
                List<Purchase> purchases = _billingClientLifecycle.purchaseUpdateEvent.getValue();
                if (purchases != null) {
                    registerPurchases(purchases);
                }
            }
        });

        fabStart = findViewById(R.id.fab_start);
        fabEnd = findViewById(R.id.fab_end);
        fabStart.setOnClickListener(new FabClickListener());
        fabEnd.setOnClickListener(new FabClickListener());
    }

    /**
     * Register SKUs and purchase tokens with the server.
     */
    private void registerPurchases(List<Purchase> purchaseList) {
        for (Purchase purchase : purchaseList) {
            String sku = purchase.getSku();
            String purchaseToken = purchase.getPurchaseToken();
            Log.d(TAG, "Register purchase with sku: " + sku + ", token: " + purchaseToken);
            _subscriptionViewModel.registerSubscription(sku, purchaseToken);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            startActivity(new Intent(this, SettingsCompatActivity.class));
            return true;
        } else if (id == R.id.sign_in) {
            startActivity(new Intent(this, GoogleSignInActivity.class));
            return true;
        } else if (id == R.id.sign_out) {

            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * Update menu based on sign-in state. Called in response to {@link #invalidateOptionsMenu}.
     */
    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        boolean isSignedIn = _authenticationViewModel.isSignedIn();
        menu.findItem(R.id.sign_in).setVisible(!isSignedIn);
        menu.findItem(R.id.sign_out).setVisible(isSignedIn);
        return true;
    }

    private void refreshData() {
        _billingClientLifecycle.queryPurchases();
        _subscriptionViewModel.manualRefresh();
    }

    /*
     * Sign in with FirebaseUI Auth.
     */
    private void triggerSignIn() {
        Log.d(TAG, "Attempting SIGN-IN!");
        List<AuthUI.IdpConfig> providers = new ArrayList<>();
        // Configure the different methods users can sign in
        providers.add(new AuthUI.IdpConfig.EmailBuilder().build());
        providers.add(new AuthUI.IdpConfig.GoogleBuilder().build());

        startActivityForResult(
                AuthUI.getInstance()
                        .createSignInIntentBuilder()
                        .setAvailableProviders(providers)
                        .build(),
                RC_SIGN_IN);
    }

    /*
     * Sign out with FirebaseUI Auth.
     */
    private void triggerSignOut() {
        _subscriptionViewModel.unregisterInstanceId();
        AuthUI.getInstance().signOut(this)
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        Log.d(TAG, "User SIGNED OUT!");
                        _authenticationViewModel.updateFirebaseUser();
                    }
                });
    }

    /*
     * Receive Activity result, including sign-in result.
     */
    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RC_SIGN_IN) {
            // If sign-in is successful, update ViewModel.
            if (resultCode == RESULT_OK) {
                Log.d(TAG, "Sign-in SUCCESS!");
                _authenticationViewModel.updateFirebaseUser();
            } else {
                Log.d(TAG, "Sign-in FAILED!");
            }
        } else {
            Log.e(TAG, "Unrecognized request code: " + requestCode);
        }
    }

    public static class SmartPagerAdapter extends SmartFragmentStatePagerAdapter {
        private final static int NUM_ITEMS = 3;

        public SmartPagerAdapter(FragmentManager fragmentManager) {
            super(fragmentManager);
        }

        // Returns total number of pages
        @Override
        public int getCount() {
            return NUM_ITEMS;
        }

        // Returns the fragment to display for that page
        @NonNull
        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    fabStart.setVisibility(View.GONE);
                    fabEnd.setVisibility(View.GONE);
                    return TabFragment1.newInstance();
                case 1:
                    fabStart.setVisibility(View.GONE);
                    fabEnd.setImageResource(R.drawable.ic_cloud_upload_white);
                    fabEnd.setVisibility(View.VISIBLE);
                    return TabFragment2.newInstance();
                case 2:
                    fabStart.setImageResource(R.drawable.ic_add_white);
                    fabStart.setVisibility(View.VISIBLE);
                    fabEnd.setImageResource(R.drawable.ic_search_white);
                    fabEnd.setVisibility(View.VISIBLE);
                    return TabFragment3.newInstance();
                default:
                    return null;
            }
        }

        // Returns the page title for the top indicator
        @Override
        public CharSequence getPageTitle(int position) {
            return "Page " + position;
        }
    }

    /*
     * Called from TabFragment1 upon tapping one of the category buttons
     */
    public void onItemSaved(Query query, String eventDate) {
        _viewPager.setCurrentItem(1); // move to tabFragment2

        Log.d(TAG, "onItemSaved() queryC="+query.getQueryC());
        Log.d(TAG, "onItemSaved() queryD="+query.getQueryD());

        try {
            ((TabFragment2) _myFragmentPagerAdapter.getItem(1))
                    .focusOnSavedItem(query, eventDate);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /*
     * Called from TabFragment3 upon tapping search button
     */
    public void onSearch(Query query, String fromDate, String toDate) {
        _viewPager.setCurrentItem(1); // move to tabFragment2

        Log.d(TAG, "onSearch() queryC="+query.getQueryC());
        Log.d(TAG, "onSearch() queryD="+query.getQueryD());

        try {
            ((TabFragment2) _myFragmentPagerAdapter.getItem(1))
                    .onSearch(query, fromDate,toDate);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    class FabClickListener implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            if (view.getId() == R.id.fab_start) {
                if (_viewPager.getCurrentItem() == 2) {
                    ((TabFragment3) _myFragmentPagerAdapter.getRegisteredFragment(2))
                            .addCriteria();
                }
            } else if (view.getId() == R.id.fab_end) {
                if (_viewPager.getCurrentItem() == 1) {
                    ((TabFragment2) _myFragmentPagerAdapter.getRegisteredFragment(1))
                            .export();
                } else if (_viewPager.getCurrentItem() == 2) {
                    ((TabFragment3) _myFragmentPagerAdapter.getRegisteredFragment(2))
                            .doSearch();
                }
            }
        }
    }
}

File: app/src/main/java/com/kakeibo/ui/MainActivity.kt
package com.kakeibo.ui

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.get
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.android.billingclient.api.Purchase
import com.bumptech.glide.Glide
import com.firebase.ui.auth.AuthUI
import com.firebase.ui.auth.AuthUI.IdpConfig
import com.firebase.ui.auth.AuthUI.IdpConfig.EmailBuilder
import com.firebase.ui.auth.AuthUI.IdpConfig.GoogleBuilder
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.navigation.NavigationView
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.google.firebase.ktx.Firebase
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.storage.ktx.storage
import com.kakeibo.Constants
import com.kakeibo.R
import com.kakeibo.SubApp
import com.kakeibo.billing.BillingClientLifecycle
import com.kakeibo.ui.model.Medium
import com.kakeibo.ui.model.Query
import com.kakeibo.ui.settings.AboutActivity
import com.kakeibo.ui.settings.InAppPurchasesActivity
import com.kakeibo.ui.settings.SettingsActivity
import com.kakeibo.ui.viewmodel.*
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    companion object {
        private const val TAG = "MainActivity"
        private const val RC_SIGN_IN = 0
        private const val RC_IN_APP_PURCHASE = 1

        lateinit var weekNames: Array<String>
        var dateFormat: Int = 0
        var numColumns: Int = 0

        private lateinit var fabStart: FloatingActionButton
        private lateinit var fabEnd: FloatingActionButton
    }

    private val firebaseStorageRef = Firebase.storage.reference

    private lateinit var _smartPagerAdapter: SmartPagerAdapter
    private lateinit var _viewPager: ViewPager2
    private lateinit var _drawerLayout: DrawerLayout
    private lateinit var _navView: NavigationView
    private lateinit var _startForResult: ActivityResultLauncher<Intent>

    private val _kkbAppViewModel: KkbAppViewModel by viewModels()
    private val _itemViewModel: ItemViewModel by viewModels()
    private val _categoryViewModel: CategoryViewModel by viewModels()
    private val _categoryDspViewModel: CategoryDspViewModel by viewModels()
    private lateinit var _billingClientLifecycle: BillingClientLifecycle
    private val _authenticationViewModel: FirebaseUserViewModel by viewModels()
    private val _billingViewModel: BillingViewModel by viewModels()
    private val _subscriptionViewModel: SubscriptionViewModel by viewModels()
    private val _medium: Medium by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        weekNames = resources.getStringArray(R.array.week_name)
        dateFormat = SubApp.getDateFormat(R.string.pref_key_date_format)
        numColumns = SubApp.getNumColumns(R.string.pref_key_num_columns)

        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        _drawerLayout = findViewById(R.id.drawer_layout)

        _navView = findViewById(R.id.nav_view)
        _navView.setNavigationItemSelectedListener(this)
        val navHeaderView = _navView.getHeaderView(0)

        val toggle = ActionBarDrawerToggle(
            this,
            _drawerLayout,
            toolbar,
            R.string.open,
            R.string.close
        )
        _drawerLayout.addDrawerListener(toggle)
        toggle.isDrawerIndicatorEnabled = true
        toggle.syncState()

        _smartPagerAdapter = SmartPagerAdapter(this)
        _viewPager = findViewById(R.id.view_pager)
        _viewPager.offscreenPageLimit = 2
        _viewPager.adapter = _smartPagerAdapter
        val tabLayout = findViewById<TabLayout>(R.id.tabs)
        TabLayoutMediator(tabLayout, _viewPager) { tab, position ->
            when(position) {
                0 -> tab.text = getString(R.string.input)
                1 -> tab.text = getString(R.string.report)
                2 -> tab.text = getString(R.string.search)
            }
        }.attach()
        tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab?) {
                when (tab!!.position) {
                    0 -> {
                        _medium.setCurrentlyShown(Medium.FRAGMENT_INPUT)
                        fabStart.visibility = View.GONE
                        fabEnd.visibility = View.GONE
                    }
                    1 -> {
                        _medium.setCurrentlyShown(Medium.FRAGMENT_REPORT)
                        fabStart.visibility = View.GONE
                        fabEnd.setImageResource(R.drawable.ic_cloud_upload_white)
                        fabEnd.visibility = View.VISIBLE
                    }
                    2 -> {
                        _medium.setCurrentlyShown(Medium.FRAGMENT_SEARCH)
                        fabStart.setImageResource(R.drawable.ic_add_white)
                        fabStart.visibility = View.VISIBLE
                        fabEnd.setImageResource(R.drawable.ic_search_white)
                        fabEnd.visibility = View.VISIBLE
                    }
                    else -> {
                    }
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab?) {
            }

            override fun onTabReselected(tab: TabLayout.Tab?) {
            }
        })

        _kkbAppViewModel.all.observe(this, {
            val showAds = it?.valInt2 == 0 // val2 = -1:original, 0:agreed to show ads

            if (showAds) {
                MobileAds.initialize(this) {}
                val adView: AdView = findViewById(R.id.ad_container)
                val adRequest = AdRequest.Builder().build()
                adView.loadAd(adRequest)
                adView.visibility = View.VISIBLE
            }
        })

//        _billingClientLifecycle = (application as SubApp).billingClientLifecycle
//        lifecycle.addObserver(_billingClientLifecycle)
//
//        /* Register purchases when they change. */
//        _billingClientLifecycle.purchaseUpdateEvent.observe(this, {
//            it?.let {
//                registerPurchases(it)
//            }
//        })
//
//        /* Launch billing flow when user clicks button to buy something. */
//        _billingViewModel.buyEvent.observe(this, {
//            it?.let {
//                _billingClientLifecycle.launchBillingFlow(this@MainActivity, it)
//            }
//        })
//
//        /* Open the Play Store when event is triggered. */
//        _billingViewModel.openPlayStoreSubscriptionsEvent.observe(this, {
//            Log.i(TAG, "Viewing subscriptions on the Google Play Store")
//            val sku = it
//            val url = if (sku == null) {
//                /* If the SKU is not specified, just open the Google Play subscriptions URL. */
//                Constants.PLAY_STORE_SUBSCRIPTION_URL
//            } else {
//                /* If the SKU is specified, open the deeplink for this SKU on Google Play. */
//                String.format(Constants.PLAY_STORE_SUBSCRIPTION_DEEPLINK_URL, sku, packageName)
//            }
//            val intent = Intent(Intent.ACTION_VIEW)
//            intent.data = Uri.parse(url)
//        })

        val profileImageView = navHeaderView.findViewById<ImageView>(R.id.imv_user_profile)
        val nameTextView = navHeaderView.findViewById<TextView>(R.id.txv_user_name)
        val emailTextView = navHeaderView.findViewById<TextView>(R.id.txv_user_email)

        _authenticationViewModel.firebaseUser.observe(this, {
            invalidateOptionsMenu()

            if (it == null) {
                Glide.with(this)
                    .load(R.mipmap.ic_launcher)
                    .circleCrop()
                    .into(profileImageView)
                nameTextView.text = getString(R.string.app_name)
                emailTextView.text = getString(R.string.not_signed_in)
                Log.d(TAG, "currently NOT logged in to firebase")
            } else {
                Glide.with(this)
                    .load(it.photoUrl)
                    .circleCrop()
                    .placeholder(R.mipmap.ic_launcher)
                    .into(profileImageView)
                nameTextView.text = it.displayName
                emailTextView.text = it.email
                Log.d(TAG, "CURRENT user: " + it.email + " " + it.displayName)
            }
        })

        /* Update subscription information when user changes. */
//        _authenticationViewModel.userChangeEvent.observe(this, {
//            _subscriptionViewModel.userChanged()
//            _billingClientLifecycle.purchaseUpdateEvent.value?.let {
//                registerPurchases(it)
//            }
//        })

//        FirebaseApp.initializeApp(this)
//        FirebaseMessaging.getInstance().subscribeToTopic(TOPIC)

//        FirebaseMessaging.getInstance().token.addOnCompleteListener(OnCompleteListener { task ->
//            if (!task.isSuccessful) {
//                Log.w(TAG, "Fetching FCM registration token failed", task.exception)
//                return@OnCompleteListener
//            }
//
//            // Get new FCM registration token
//            val token = task.result
//
//            // Log and toast
//            val msg = getString(R.string.msg_token_fmt, token)
//            Log.d(TAG, msg)
//            Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
//        })

        _startForResult = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) { result: ActivityResult ->
            if (result.resultCode == Activity.RESULT_OK) {
                Toast.makeText(this, R.string.sign_in_success, Toast.LENGTH_LONG).show()
                _authenticationViewModel.updateFirebaseUser()
            } else {
                Toast.makeText(this, R.string.sign_in_failure, Toast.LENGTH_LONG).show()
            }
        }

        fabStart = findViewById(R.id.fab_start)
        fabEnd = findViewById(R.id.fab_end)
        fabStart.setOnClickListener(FabClickListener())
        fabEnd.setOnClickListener(FabClickListener())
    }

    override fun onResume() {
        super.onResume()
        refreshData()
    }

    override fun onBackPressed() {
        if (_viewPager.currentItem <= 3) {
            // If the user is currently looking at the first step, allow the system to handle the
            // Back button. This calls finish() on this activity and pops the back stack.
            super.onBackPressed()
        } else {
            // Otherwise, select the previous step.
            _viewPager.currentItem = _viewPager.currentItem - 1
        }
    }

    /*
     * Register SKUs and purchase tokens with the server.
     */
//    private fun registerPurchases(purchaseList: List<Purchase>) {
//        val navHeaderView = _navView.getHeaderView(0)
//        val subsTextView = navHeaderView.findViewById<TextView>(R.id.txv_subscription)
//        if (purchaseList.isEmpty()) {
//            subsTextView.visibility = View.GONE
//        }
//
//        for (purchase in purchaseList) {
//            val sku = purchase.skus[0]
//            val purchaseToken = purchase.purchaseToken
//            Log.d(TAG, "Register purchase with sku: $sku, token: $purchaseToken")
//
//            _subscriptionViewModel.registerSubscription(sku, purchaseToken)
//
//            subsTextView.visibility = View.VISIBLE
//            if (sku == Constants.BASIC_SKU) subsTextView.text = "Subscribed: Basic Plan" //todo translation
//            if (sku == Constants.PREMIUM_SKU) subsTextView.text = "Subscribed: Premium Plan"
//        }
//    }

    private fun refreshData() {
        _billingClientLifecycle.queryPurchases()
        _subscriptionViewModel.manualRefresh()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)

        val isSignedIn = _authenticationViewModel.isSignedIn()
        _navView.menu[0].isVisible = !isSignedIn
        _navView.menu[1].isVisible = isSignedIn

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.sign_in -> {
//                startActivity(Intent(this, SignInActivity::class.java))
                triggerSignIn()
                true
            }
            R.id.sign_out -> {
                triggerSignOut()
                true
            }
//            R.id.in_app_purchases -> {
//                _startForResult.launch(Intent(this, InAppPurchasesActivity::class.java))
//                true
//            }
            R.id.about -> {
                startActivity(Intent(this, AboutActivity::class.java))
                true
            }
            else -> true
        }
    }

    /*
     * Sign in with FirebaseUI Auth.
     */
    private fun triggerSignIn() {
        Log.d(TAG, "Attempting SIGN-IN!")
        val providers: MutableList<IdpConfig> = ArrayList()
        providers.add(EmailBuilder().build())
        providers.add(GoogleBuilder().build())
        val intent = AuthUI.getInstance()
            .createSignInIntentBuilder()
            .setAvailableProviders(providers)
            .build()
        _startForResult.launch(intent)
    }

    /*
     * Sign out with FirebaseUI Auth.
     */
    private fun triggerSignOut() {
        _subscriptionViewModel.unregisterInstanceId()
        AuthUI.getInstance().signOut(this).addOnCompleteListener {
            Log.d(TAG, "User SIGNED OUT!")
            _authenticationViewModel.updateFirebaseUser()
            Toast.makeText(this, R.string.sign_out_success, Toast.LENGTH_LONG).show()
        }
    }

    class SmartPagerAdapter(fa: FragmentActivity) : FragmentStateAdapter(fa) {
        val fragments: MutableList<Fragment> = mutableListOf()

        override fun getItemCount(): Int = 3

        override fun createFragment(position: Int): Fragment {
            return when (position) {
                0 -> {
                    val fragment1 = InputFragment.newInstance()
                    fragments.add(fragment1)
                    fragment1
                }
                1 -> {
                    val fragment2 = ReportFragment.newInstance()
                    fragments.add(fragment2)
                    fragment2
                }
                2 -> {
                    val fragment3 = SearchFragment.newInstance()
                    fragments.add(fragment3)
                    fragment3
                }
                else -> throw Exception("unknown item type")
            }
        }
    }

    /*
     * Called from TabFragment1 upon tapping one of the category buttons
     */
    fun onItemSaved(date: String) {
        _viewPager.currentItem = 1 // move to tabFragment2
        if (_smartPagerAdapter.fragments.size == 0) {
            _smartPagerAdapter.createFragment(0)
            _smartPagerAdapter.createFragment(1)
            _smartPagerAdapter.createFragment(2)
        }
        (_smartPagerAdapter.fragments[1] as ReportFragment).focusOnSavedItem(date)
    }

    /*
     * Called from TabFragment3 upon tapping search button
     */
    fun onSearch(query: Query) {
        _viewPager.currentItem = 1 // move to tabFragment2
        if (_smartPagerAdapter.fragments.size == 0) {
            _smartPagerAdapter.createFragment(0)
            _smartPagerAdapter.createFragment(1)
            _smartPagerAdapter.createFragment(2)
        }
        (_smartPagerAdapter.fragments[1] as ReportFragment).onSearch(query)
    }

    internal inner class FabClickListener : View.OnClickListener {
        override fun onClick(view: View) {
            if (view.id == R.id.fab_start) {
                if (_viewPager.currentItem == 2) {
                    (_smartPagerAdapter.fragments[2] as SearchFragment).addCriteria()
                }
            }
            else if (view.id == R.id.fab_end) {
                if (_viewPager.currentItem == 1) {
                    (_smartPagerAdapter.fragments[1] as ReportFragment).export()
                } else if (_viewPager.currentItem == 2) {
                    (_smartPagerAdapter.fragments[2] as SearchFragment).doSearch()
                }
            }
        }
    }
}